src/lib/index/wb_cache.cpp

```diff
@@ -431,32 +485,103 @@ void IndexWBCache::recover(sisl::byte_view sb) {
                         m_vdev->commit_blk(buf->m_blkid);
                     }
                     l0_bufs.push_back(buf);
+                    // insert upbuffer and its upbuffer to the recovering_upbuffers set until nullptr
+                    // suppose the following scenario that crash happened when flushing B
+                    //    ├── A (WAITING FOR B)
+                    //    │    ├── B (CRASHED)
+                    //    │    │    ├── C (FLUSHED)
+                    //    │    │    └── D (FLUSHED)
+                    //    │    │         └── E (FLUSHED)
+                    //    │    └── F (FLUSHED)
+                    //    └── G (FLUSHED)
+                    // we need to recover B and A. During recover_buf(A), we need to recover F and commit its blk. So in
+                    // this case, we give a second chance to new buffers to be a part of repair. There is still a case
+                    // that committing new buffers is not needed since they are not part of recovery paths.
+                    auto up_buf = buf->m_up_buffer;
+                    while (up_buf) {
+                        recovering_upbuffers.insert(up_buf);
```

Comment:
We can break here if the insertion fails to skip duplicate loops.

Reply:
it is a set